src/Microsoft.OData.Edm/Csdl/Parsing/SchemaJsonParser.cs

```diff
@@ -0,0 +1,1564 @@
+            {
+                // entitySet
+                if (!isCollection.Value)
+                {
+                    // If presented, $IsCollection should be true
+                    context.ReportError(EdmErrorCode.JsonInvalidCollectionValue, "$IsCollection should be true");
+                }
+
+                if (includeInServiceDocument == null || includeInServiceDocument.Value)
+                {
+                    navigationSource = new CsdlEntitySet(name, type, navigationPropertyBindings, location);
+                }
+                else
+                {
+                    navigationSource = new CsdlEntitySet(name, type, navigationPropertyBindings, location, false);
+                }
+            }
+            else
+            {
+                // singleton
+                navigationSource = new CsdlSingleton(name, type, navigationPropertyBindings, location);
+            }
+
+            foreach (var annotation in annotations)
+            {
+                navigationSource.AddAnnotation(annotation);
+            }
+
+            return navigationSource != null;
+        }
+
+        public static IList<CsdlNavigationPropertyBinding> ParseCsdlNavigationPropertyBinding(JsonElement element, JsonParserContext context)
+        {
+            if (element.ValueKind != JsonValueKind.Object)
+            {
+                return null;
+            }
+
+            CsdlLocation location = context.Location();
+
+            IList<CsdlNavigationPropertyBinding> bindings = new List<CsdlNavigationPropertyBinding>();
+            foreach (JsonProperty property in element.EnumerateObject())
+            {
+                string bindingPath = property.Name;
+                string target = property.Value.GetString();
+
+                bindings.Add(new CsdlNavigationPropertyBinding(bindingPath, target, location));
+            }
+
+            return bindings;
+        }
+        #endregion
+
+        #region Structured type
+        /// <summary>
+        /// Parse the input <see cref="JsonElement"/> to the <see cref="CsdlEntityType"/>.
+        /// </summary>
+        /// <param name="name">the unqualified name of the entity type.</param>
+        /// <param name="entityObject">The json value to parse from.</param>
+        /// <param name="jsonPath">The input json value path.</param>
+        /// <returns>null or the parsed <see cref="CsdlEntityType"/>.</returns>
+        internal static CsdlEntityType ParseCsdlEntityType(string name, JsonElement entityObject, JsonParserContext context)
+        {
+            // An entity type is represented as a member of the schema object whose name is the unqualified name of the entity type and whose value is an object.
+            Debug.Assert(entityObject.ValueKind == JsonValueKind.Object);
+
+            // The entity type object MUST contain the member $Kind with a string value of EntityType.
+            ValidateKind(entityObject, context, "EntityType", required: true);

```

Comment:
I don't think that "$kind" is used for entity sets any longer. Please make sure it's neither written nor read.